File: scilab/routines/system/terme.f
      subroutine terme
c
c     Copyright INRIA
      include '../stack.h'
      integer r,op,bslash,star,slash,dot,not
      integer plus,minus
c      integer equal,less,great,not

      data bslash/49/,star/47/,slash/48/,dot/51/,not/61/
      data plus/45/,minus/46/

c
      r = rstk(pt)
      if (ddt .eq. 4) then
         write(buf(1:8),'(2i4)') pt,r
         call basout(io,wte,' term   pt:'//buf(1:4)//' rstk(pt):'
     &               //buf(5:8))
      endif
c
      ir=r/100
      if(ir.ne.2) goto 01
      goto(05,25,26,25),r-200
      goto 99
c
c
c     premier facteur
c-------------------
   01 pt = pt+1
      rstk(pt) = 201
      icall=3
c     *call* factor
      return
   05 pt = pt-1
   10 op = 0
      if (sym .eq. dot) then
         op = dot
         call getsym
      endif
      if (sym.eq.star .or. sym.eq.slash .or. sym.eq.bslash ) go to 20
      if (op.ne.0) then
         call error(7)
         return
      endif
      return
c
c     facteurs suivants
c----------------------
   20 op = op + sym
      call getsym
      if (sym .eq. dot) op = op + 2*sym
      if (sym .eq. dot) call getsym
      pt = pt+1
      pstk(pt) = op
      if (sym .ne. not) goto 21
      rstk(pt)=204
      icall=1
c     *call* lfact (in expr)
      return

 21   rstk(pt) = 202
      if(sym.eq.plus.or.sym.eq.minus) then
c     .  added to handle syntax like a*-b for Matlab compatiblity
         icall=1
c     .  *call* expr
      else
         icall=3
c     .  *call* factor
      endif
      return
   25 fin=pstk(pt)
c     evaluation
c---------------
      rstk(pt)=203
      rhs=2
      icall=4
c     *call* allops(op)
      return
   26 pt=pt-1
      go to 10
c
   99 call error(22)
      if (err .gt. 0) return
      return
      end

File: scilab/modules/core/src/c/terme.c
/*
 * Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
 * Copyright (C) 2007 - INRIA - Allan CORNET
 * 
 * This file must be used under the terms of the CeCILL.
 * This source file is licensed as described in the file COPYING, which
 * you should have received as part of this distribution.  The terms
 * are also available at    
 * http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt
 *
 */
#include <string.h>
#include <stdio.h>
#include "terme.h"
#include "stack-def.h"
#include "stack-c.h"
#include "basout.h"
#include "do_error_number.h"
/*--------------------------------------------------------------------------*/ 
extern int C2F(getsym)();
/*--------------------------------------------------------------------------*/ 
int C2F(terme)(void)
{
#define char_plus 45
#define char_minus 46
#define char_bchar_slash 49
#define char_star 47
#define char_slash 48
#define char_dot 51
#define char_not 61
#define constnumber 114688

  /* Local variables */
  int op = 0;

  int r = 0;

  /* int equal,less,great,char_not */
  r = C2F(recu).rstk[(constnumber + (0 + ( (C2F(recu).pt - 1) << 2)) - constnumber) / 4];

  if (C2F(iop).ddt == 4) {
    static char tmp[100];
    static int io;
    sprintf(tmp," terme pt:%d rstk(pt):%d sym:%d",C2F(recu).pt, C2F(recu).rstk[C2F(recu).pt - 1], C2F(com).sym);
    C2F(basout)(&io, &C2F(iop).wte,tmp, (long)strlen(tmp));

  }

  if ( (r / 100) != 2) 
    { /* first factor */
      ++C2F(recu).pt;

      /* escape to call fact */
      C2F(recu).rstk[C2F(recu).pt - 1] = 201;
      C2F(recu).icall = 3;
      return 0;
    }

  switch (r - 200) 
    {
    case 1: 
      { /* return point after  first factor evaluation*/
	--C2F(recu).pt;
	op = 0;
	if (C2F(com).sym == char_dot) 
	  {
	    op = char_dot;
	    C2F(getsym)();
	  }

	if (C2F(com).sym == char_star || C2F(com).sym == char_slash || C2F(com).sym == char_bchar_slash) 
	  {
	    op += C2F(com).sym;
	    C2F(getsym)();
	    if (C2F(com).sym == char_dot) op += C2F(com).sym << 1;

	    if (C2F(com).sym == char_dot) C2F(getsym)();

	    ++C2F(recu).pt;
	    C2F(recu).pstk[C2F(recu).pt - 1] = op;
	    if (C2F(com).sym != char_not) 
	      {
		/* escape to call fact */
		C2F(recu).rstk[C2F(recu).pt - 1] = 202;
		C2F(recu).icall = 3;
		return 0;
	      }
	   
	    /* escape to call expr */
	    C2F(recu).rstk[C2F(recu).pt - 1] = 204;
	    C2F(recu).icall = 1;
	    return 0;
	  }

	if (op != 0) 
	  {
	    SciError(7);
	  }
	return 0;
      }
    case 2:
      { /* return point after n th factor evaluation*/
	Fin = C2F(recu).pstk[C2F(recu).pt - 1];
	/* evaluation */
	Rhs = 2;

	/* escape to call allops(op) */
	C2F(recu).rstk[C2F(recu).pt - 1] = 203;
	C2F(recu).icall = 4;
	return 0;
      }
    case 3: 
      {  /* return point after operation evaluation (allops(op))*/
	--C2F(recu).pt;
	op = 0;
	if (C2F(com).sym == char_dot) 
	  {
	    op = char_dot;
	    C2F(getsym)();
	  }
	if (C2F(com).sym == char_star || C2F(com).sym == char_slash || C2F(com).sym == char_bchar_slash) 
	  {
	    op += C2F(com).sym;
	    C2F(getsym)();
	    if (C2F(com).sym == char_dot) op += C2F(com).sym << 1;

	    if (C2F(com).sym == char_dot) C2F(getsym)();

	    ++C2F(recu).pt;
	    C2F(recu).pstk[C2F(recu).pt - 1] = op;
	    if (C2F(com).sym != char_not) 
	      {
		C2F(recu).rstk[C2F(recu).pt - 1] = 202;
		/*next line added to handle syntax like a*-b for Matlab compatiblity */
		if (C2F(com).sym == char_plus || C2F(com).sym == char_minus) 
		  C2F(recu).icall = 1; /* escape to call expr */
		else
		  C2F(recu).icall = 3; /* escape to call fact */
		return 0;
	      }

	    /* escape to evaluate a logical factor (in expr) */
	    C2F(recu).rstk[C2F(recu).pt - 1] = 204;
	    C2F(recu).icall = 1;
	    return 0;
	  }

	if (op != 0) 
	  {
	    SciError(7);
	  }
	return 0;
      }
    case 4: 
      { /* return point after logical factor evaluation (in expr)*/ 
	Fin = C2F(recu).pstk[C2F(recu).pt - 1];
	Rhs = 2;

	/* escape to call allops(op) */
	C2F(recu).rstk[C2F(recu).pt - 1] = 203;
	C2F(recu).icall = 4;
	return 0;
      }
      break;
    }

  SciError(22);
  return 0;
}
/*--------------------------------------------------------------------------*/ 
